File: services/nuldoc/nuldoc-src/generators/post.ts
import { join } from "@std/path";
import PostPage from "../pages/PostPage.ts";
import { Config } from "../config.ts";
import { Document } from "../markdown/document.ts";
import { Page } from "../page.ts";
import { Date, Revision } from "../revision.ts";

export interface PostPage extends Page {
  title: string;
  description: string;
  tags: string[];
  revisions: Revision[];
  published: Date;
  updated: Date;
  uuid: string;
}

export function getPostPublishedDate(page: { revisions: Revision[] }): Date {
  for (const rev of page.revisions) {
    if (!rev.isInternal) {
      return rev.date;
    }
  }
  return page.revisions[0].date;
}

export function getPostUpdatedDate(page: { revisions: Revision[] }): Date {
  return page.revisions[page.revisions.length - 1].date;
}

export function postHasAnyUpdates(page: { revisions: Revision[] }): boolean {
  return 2 <= page.revisions.filter((rev) => !rev.isInternal).length;
}

export async function generatePostPage(
  doc: Document,
  config: Config,
): Promise<PostPage> {
  const html = await PostPage(doc, config);

  const cwd = Deno.cwd();
  const contentDir = join(cwd, config.locations.contentDir);
  const destFilePath = join(
    doc.sourceFilePath.replace(contentDir, "").replace(".md", ""),
    "index.html",
  );
  return {
    root: html,
    renderer: "html",
    site: "blog",
    destFilePath: destFilePath,
    href: destFilePath.replace("index.html", ""),
    title: doc.title,
    description: doc.description,
    tags: doc.tags,
    revisions: doc.revisions,
    published: getPostPublishedDate(doc),
    updated: getPostUpdatedDate(doc),
    uuid: doc.uuid,
  };
}

File: services/nuldoc/lib/nuldoc/generators/post.rb
module Nuldoc
  PostPage = Data.define(:root, :renderer, :site, :dest_file_path, :href,
                         :title, :description, :tags, :revisions, :published, :updated,
                         :uuid, :source_file_path)

  class GeneratorUtils
    def self.published_date(page)
      page.revisions.each do |rev|
        return rev.date unless rev.is_internal
      end
      page.revisions[0].date
    end

    def self.updated_date(page)
      page.revisions.last.date
    end

    def self.any_updates?(page)
      page.revisions.count { |rev| !rev.is_internal } >= 2
    end
  end

  module Generators
    class Post
      def initialize(doc, config)
        @doc = doc
        @config = config
      end

      def generate
        html = Pages::PostPage.new(doc: @doc, config: @config).render

        content_dir = File.join(Dir.pwd, @config.locations.content_dir)
        dest_file_path = File.join(
          @doc.source_file_path.sub(content_dir, '').sub('.md', ''),
          'index.html'
        )

        PostPage.new(
          root: html,
          renderer: :html,
          site: 'blog',
          dest_file_path: dest_file_path,
          href: dest_file_path.sub('index.html', ''),
          title: @doc.title,
          description: @doc.description,
          tags: @doc.tags,
          revisions: @doc.revisions,
          published: GeneratorUtils.published_date(@doc),
          updated: GeneratorUtils.updated_date(@doc),
          uuid: @doc.uuid,
          source_file_path: @doc.source_file_path
        )
      end
    end
  end
end
